src/rendezvous/channels/ecdhV1.ts

```diff
@@ -0,0 +1,298 @@
+        'raw',
+        key,
+        { name: 'AES-GCM' },
+        false,
+        ['encrypt', 'decrypt'],
+    );
+
+    return imported;
+}
+
+/**
+ * Implementation of the unstable [MSC3903](https://github.com/matrix-org/matrix-spec-proposals/pull/3903)
+ * X25519/ECDH key agreement based secure rendezvous channel.
+ * Note that this is UNSTABLE and may have breaking changes without notice.
+ */
+export class MSC3903ECDHv1RendezvousChannel implements RendezvousChannel {
+    private olmSAS?: SAS;
+    private ourPublicKey: Uint8Array;
+    private aesKey?: CryptoKey | Uint8Array;
+
+    constructor(
+        public transport: RendezvousTransport,
+        private theirPublicKey?: Uint8Array,
+        public onFailure?: (reason: RendezvousFailureReason) => void,
+    ) {
+        this.olmSAS = new global.Olm.SAS();
+        this.ourPublicKey = decodeBase64(this.olmSAS.get_pubkey());
+    }
+
+    public async generateCode(intent: RendezvousIntent): Promise<ECDHv1RendezvousCode> {
+        if (this.transport.ready) {
+            throw new Error('Code already generated');
+        }
+
+        await this.send({ algorithm: SecureRendezvousChannelAlgorithm.ECDH_V1 });
+
+        const rendezvous: ECDHv1RendezvousCode = {
+            "rendezvous": {
+                algorithm: SecureRendezvousChannelAlgorithm.ECDH_V1,
+                key: encodeBase64(this.ourPublicKey),
+                transport: await this.transport.details(),
+            },
+            intent,
+        };
+
+        return rendezvous;
+    }
+
+    async connect(): Promise<string> {
+        if (!this.olmSAS) {
+            throw new Error('Channel closed');
+        }
+
+        const isInitiator = !this.theirPublicKey;
+
+        if (isInitiator) {
+            // wait for the other side to send us their public key
+            const res = await this.receive();
+            if (!res) {
+                throw new Error('No response from other device');
+            }
+            const { key, algorithm } = res;
+
+            if (algorithm !== SecureRendezvousChannelAlgorithm.ECDH_V1 || !key) {
+                throw new RendezvousError(
+                    'Unsupported algorithm: ' + algorithm,
+                    RendezvousFailureReason.UnsupportedAlgorithm,
+                );
+            }
+
+            this.theirPublicKey = decodeBase64(key);
+        } else {
+            // send our public key unencrypted
+            await this.send({
+                algorithm: SecureRendezvousChannelAlgorithm.ECDH_V1,
+                key: encodeBase64(this.ourPublicKey),
+            });
+        }
+
+        this.olmSAS.set_their_key(encodeBase64(this.theirPublicKey!));
+
+        const initiatorKey = isInitiator ? this.ourPublicKey : this.theirPublicKey!;
+        const recipientKey = isInitiator ? this.theirPublicKey! : this.ourPublicKey;
+        let aesInfo = SecureRendezvousChannelAlgorithm.ECDH_V1.toString();
+        aesInfo += `|${encodeBase64(initiatorKey)}`;
```

Comment:
I'd suggest using unpadded base64 encoding in general for the final specced version, but this may be OK for now (and of course changing it here would require changing it in the other implementations)

Reply:
Yeah, it's a bit late. Please do feedback into the actual MSCs and I can update.